### ISSUE TITLE
Dimension mismatch error when running mxnetR

### ISSUE THREAD

**Issue by nnguyen24** [dmlc/mxnet#6596]
I am trying to build a Feed Forward neural network with MXNetR. My input is a data frame with 6380 rows and 180 columns. My training and testing outputs are one-dimensional vectors with 319 elements each.

I run the model with the batch size set to 1 and the number of neurons at the output layer set to 319. So for each batch, I expected to get a vector with 319 elements. I aim to minimize my loss function, which is the correlation between my predicted output vector and actual output vector.

Below is my code:
```
    # Generate testing data
    train.x = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 6380)
    test.x = matrix(data = rexp(200, rate = 10), nrow = 60, ncol = 6380)
    train.y = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 319)
    test.y = matrix(data = rexp(200, rate = 10), nrow = 60, ncol = 319)
   
    # Reshape testing data
   train.array <-train.x
   dim(train.array) <-c(20,319,1,ncol(train.x))
   test.array<-test.x
   dim(test.array) <-c (20,319,1,ncol(test.x))

    # Define the input data
    data <- mx.symbol.Variable("data")
    
    # Define the first fully connected layer
    fc1 <- mx.symbol.FullyConnected(data, num_hidden = 100)
    act.fun <- mx.symbol.Activation(fc1, act_type = "relu") # create a hidden layer with Rectified Linear Unit as its activation function.
    output <<- mx.symbol.FullyConnected(act.fun, num_hidden = 319)
    
    # Customize loss function
    label <- mx.symbol.Variable("label")
    lro <- mx.symbol.MakeLoss(mx.symbol.Correlation(mx.symbol.reshape(output 
    ,shape = c(1,319)),mx.symbol.reshape(label, shape = c(1,319,1,1))))

    model <- mx.model.FeedForward.create(symbol=lro, X=train.x, y=train.y, 
                                     num.round=5000, array.batch.size=1, optimizer = "adam",
                                     learning.rate = 0.0003, eval.metric = mx.metric.rmse,
                                     epoch.end.callback = mx.callback.log.train.metric(20, logger))
```

And here is the error when I run the code above:

```
[Error in symbol$infer.shape(list(...)) : 
Error in operator correlation6: [14:36:37] src/operator/./correlation-inl.h:177: Check failed: 
 dshape2.ndim() == 4U (0 vs. 4) data should be a 4D tensor

Stack trace returned 10 entries:
[bt] (0) /usr/lib64/R/library/mxnet/libs/libmxnet.so(_ZN4dmlc15LogMessageFatalD1Ev+0x29) [0x7f64602b28b9]
[bt] (1) /usr/lib64/R/library/mxnet/libs/libmxnet.so(_ZNK5mxnet2op15CorrelationProp10InferShapeEPSt6vectorIN4nnvm6TShapeESaIS4_EES7_S7_+0x33d) [0x7f6460f082bd]
[bt] (2) /usr/lib64/R/library/mxnet/libs/libmxnet.so(+0xd461f9) [0x7f6460d841f9]
[bt] (3) /usr/lib64/R/library/mxnet/libs/libmxnet.so(+0x116630f) [0x7f64611a430f]
[bt] (4) /usr/lib64/R/library/mxnet/libs/libmxnet.so(+0x1167bb2) [0x7f64611a5bb2]
[bt] (5) /usr/lib64/R/library/mxnet/libs/libmxnet.so(_ZN4nnvm11ApplyPassesENS_5GraphERKSt6vectorISsSaISsEE+0x501) [0x7f64611c1481]
[bt] (6) /usr/lib64/R/library/mxnet/libs/libmxnet.so(_ZN4nnvm9ApplyPassENS_5GraphERKSs+0x8e) [0x7f64610f9f2e]
[bt] (7) /usr/lib64/R/library/mxnet/libs/libmxnet.s
```
At the moment, I am clueless about how I should fix this error. I have been searching for a way to reshape my data sets so that they are 4D tensors but could not find any. I do not look for an explicit solution for my problem, but any suggestions on how I should tackle this error would be greatly appreciated.

**Comment by offerm** [comment]
This was working for me one I added the missing "c"  for the shape:

lro <- mx.symbol.MakeLoss(mx.symbol.Correlation(mx.symbol.reshape(output,shape = c(1,319)),label))

**Comment by offerm** [comment]
You also to add X and y and where you get the error.
Make sure you have a reproducible code.

**Comment by offerm** [comment]
Hard to say without the and X example.

It might also be due to batch.size=1.

**Comment by nnguyen24** [comment]
@offerm: I updated the code to make sure that it includes a test data set. Will try to experiment with different batch sizes.

**Comment by offerm** [comment]
I simplified the code so it will be easier to understand the situation (consider to change your example above):
# Generate testing data
train.x = matrix(data = rexp(120*6380, rate = 10), nrow = 120, ncol = 6380)
train.y = matrix(data = rexp(120*319, rate = 10), nrow = 120, ncol = 319)

# Define the input data
data <- mx.symbol.Variable("data")
# Define the first fully connected layer
fc1 <- mx.symbol.FullyConnected(data, num_hidden = 319)
# Customize loss function
label <- mx.symbol.Variable("label")
lro <- mx.symbol.MakeLoss(mx.symbol.Correlation(fc1 ,label))

model <- mx.model.FeedForward.create(symbol=lro, 
                                     X=t(train.x), 
                                     y=t(train.x), 
                                     num.round=5000, array.batch.size=1, optimizer = "adam",
                                     learning.rate = 0.0003, 
                                     #eval.metric = mx.metric.rmse,
                                     array.layout = "colmajor",
                                     epoch.end.callback = mx.callback.log.train.metric(20, logger))

With this, you get your original error:

[16:13:04] /Users/admin/github/mxnet/dmlc-core/include/dmlc/logging.h:304: [16:13:04] src/operator/./correlation-inl.h:176: Check failed: dshape1.ndim() == 4U (2 vs. 4) data should be a 4D tensor

The error is coming directly from the correlation c function that checks the shape for the arguments. I still try to understand the C code there.

I tried to simplified even more by using executor code as below:

require(mxnet)
A = mx.symbol.Variable('A')
B = mx.symbol.Variable('B')
#C = mx.symbol.Correlation(A,B)
C = mx.symbol.sort(A+B)
a = mx.nd.zeros(c(4), mx.cpu())
b = mx.nd.array(as.array(c(7, 2,4,5)), mx.cpu())

exec = mxnet:::mx.symbol.bind(
  symbol=C,
  ctx=mx.cpu(),
  arg.arrays = list(A=a, B=b),
  aux.arrays = list(),
  grad.reqs = list("null", "null"))

# calculate outputs
mx.exec.forward(exec)
out = as.array(exec$outputs[[1]])
print(out)

It is working as expected until you change the mx.symbol.sort to mx.symbol.correlation (A,B).

C = mx.symbol.Correlation(A,B)


When you do and run it, it crashes.....

Will try to give it some more time over the weekend.

**Comment by nnguyen24** [comment]
@offerm: Thank you. I will spend time looking at the problem myself as well.

**Comment by offerm** [comment]
Also check test_operator.py line 1721 (unittest_correlation). Looks like it does expect 4U tensor.

**Comment by offerm** [comment]
So  I managed to use the correlation with this code:

require(mxnet)
A = mx.symbol.Variable('A')
B = mx.symbol.Variable('B')
C = mx.symbol.Correlation(A,B,kernel_size = 1,max_displacement = 1,stride1 = 1,stride2 = 1
                          ,pad_size = 4,is_multiply = FALSE)
#C = mx.symbol.sort(A+B)
a = mx.nd.reshape(mx.nd.zeros(c(4), mx.cpu()),shape=c(1,1,1,4))
b = mx.nd.reshape(mx.nd.array(as.array(c(7, 2,4,5)), mx.cpu()),shape=c(1,1,1,4))

exec = mxnet:::mx.symbol.bind(
  symbol=C,
  ctx=mx.cpu(),
  arg.arrays = list(A=a, B=b),
  aux.arrays = list(),
  grad.reqs = list("null", "null"))

# calculate outputs
mx.exec.forward(exec)
out = as.array(exec$outputs[[1]])
print(out)


However, as you can see from the result this is not a simple correlation of two arrays.

**Comment by thirdwing** [comment]
@nnguyen24 You need to make the two symbols in `mx.symbol.Correlation` have exactly the same shape:

```r
library(mxnet)

train.x = matrix(data = rexp(120 * 319, rate = 10), nrow = 120, ncol = 319)

# Define the input data
data <- mx.symbol.Variable("data")
# Define the first fully connected layer
fc1 <- mx.symbol.FullyConnected(data, num_hidden = 100)
act.fun <- mx.symbol.Activation(fc1, act_type = "relu") # create a hidden layer with Rectified Linear Unit as its activation function.
output <- mx.symbol.FullyConnected(act.fun, num_hidden = 319)
# Customize loss function
label <- mx.symbol.Variable("label")
lro <- mx.symbol.MakeLoss(mx.symbol.Correlation(
  mx.symbol.reshape(output, shape = c(100, 319, 1, 1)),
  mx.symbol.reshape(label, shape = c(100, 319, 1, 1))))

model <- mx.model.FeedForward.create(symbol = lro, X = t(train.x),
                                     y = t(train.x), num.round = 5,
                                     array.batch.size = 100,
                                     optimizer = "sgd",
                                     learning.rate = 0.0003,
                                     array.layout = "colmajor",
                                     epoch.end.callback = mx.callback.log.train.metric(20, logger))
```
